Add organizer debrief survey

outdoor2kode opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here. Each issue needs the "fixes" keyword if the PR fixes more than one thing. -->
Closes #930

This PR implements what has been concluded in the ticket. The logic refers to #997.

**Flow**

- Manually published Organizer Survey (event debrief) in Central.

- When the plugin is activated:
  - Create a `tix_email` in draft status.
    -  The email will contain a token generated by hashing the access token key and WordCamp ID, to prevent the survey page from being accessed and abused by individuals other than the lead organizers.
  - Schedule a `daily` cron job.

- When the cron job runs:
  - Check if we already sent an email for this camp.
  - Check to make sure the camp ended exactly 2 days ago.
  - Add the lead organizer as an attendee and associate it with `tix_email`.
  - Add `tix_email` to the queue (set to `pending`)
    - _email gets picked up by camptix cron email sending queue `tix_scheduled_every_ten_minutes`_

## Screenshots

| View | Screenshot |
| --- | --- |
| Email queue | <img width="1525" alt="image" src="https://github.com/WordPress/wordcamp.org/assets/18050944/36c72cc4-7e74-46ea-9a28-446c525a6864"> |
| Manually published Organizer Survey (event debrief) page in central | <img width="1528" alt="image" src="https://github.com/WordPress/wordcamp.org/assets/18050944/28bd62d4-5cf8-47c9-b990-9a8f7a4e9572"> |
| Draft Survey Email | <img width="1234" alt="image" src="https://github.com/WordPress/wordcamp.org/assets/18050944/75fc0963-b93f-46bb-a18a-f0e19ec4a05a"> |
| Survey Email Sent | <img width="608" alt="image" src="https://github.com/WordPress/wordcamp.org/assets/18050944/3d3d3165-8183-458e-ad78-85508545ea17"> |
| Survey | <img width="821" alt="image" src="https://github.com/WordPress/wordcamp.org/assets/18050944/abbe8b65-d240-4789-bb6e-b1a5bb3c4c53"> |

## How to test

1. Create a Next-Gen Event WordCamp site and fill in `Organizer Name` and `Email Address` inside the Organizing Team.
   - Make the Start Date in the past.
   - Make the End Date the current date.
2. Visit Cenral
   - Go to Pages
     - Manually publish `Organizer Survey (event debrief)`
     - View the content, which should be the same as the description in ticket.
2. Visit Next-Gen
   - Go to Ticket -> E-Mail
     - Expect to see an email `Organizer Survey (event debrief)` in `Draft` mode.
     - The email content is pending an update from the Tardis Team. 
   - Run the following wp cli script to make sure the cron jobs are scheduled
     - `wp cron event list --url=https://events.wordpress.test/{city}/{year}/{type}` 
     - expect to see the following cron job: `wc_organizer_debrief_survey_email`
   - Go back to the Next-Gen Wordcamp Post type and update the End Date to exact 2 days ago.
   - Run this script to set WordCamp to closed `wp post update {post_id} --post_status=wcpt-closed --url=https://central.wordcamp.test/`
   - Run the following cli command:
     - run `wp cron event run wc_organizer_debrief_survey_email --url=https://events.wordpress.test/{city}/{year}/{type}` 
       - Should succeed
       - Status of `Organizer Survey (event debrief)` in Ticket -> E-Mail should be `pending`.
       - An attendee should be added in Ticket -> Attendees with `tix_first_name` and `tix_email` the same as the one you input when creating the event.
     - run `wp cron event run tix_scheduled_every_ten_minutes -url=https://events.wordpress.test/{city}/{year}/{type}`
       - `Organizer Survey (event debrief)` in Ticket -> E-Mail should be published.
       - In, http://localhost:1080/, there should be an email with Subject `Organizer Survey (event debrief)`. 
         - Visit the survey link on the front end and submit.
 4. Make sure none of this exists on the traditional WordCamp.

outdoor2kode committed:
Add plugin: WordCamp Organizer Survey.

outdoor2kode committed:
Add cron job basic framework.

outdoor2kode committed:
Add email basic framework.

outdoor2kode committed:
Add central admin-page.

outdoor2kode committed:
Add debrief survey page.

outdoor2kode committed:
Integrate email with Plugin: WordCamp Organizer Reminders

outdoor2kode committed:
Revert Integrating email with Organizer Reminders.

The Organizer Reminders only works on central. It would get too complicated and introduce
a few glitches if trying to use it through event site. The extra work doesn't worth the
effort at the moment.

outdoor2kode committed:
Update cron job.

outdoor2kode committed:
Tidy up file structure

outdoor2kode committed:
Naming tidy up.

renintw commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/admin-page.php:
```diff
@@ -8,9 +8,8 @@
 defined( 'WPINC' ) || die();
```
This commit is reverted, can be skipped.

outdoor2kode commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/admin-page.php:
```diff
@@ -8,9 +8,8 @@
 
 use function WordCamp\OrganizerSurvey\{get_feature_id};
-// use function WordCamp\OrganizerSurvey\Email\{get_email_id};
+use function WordCamp\OrganizerSurvey\Reminder\{get_reminder_id};
```
This commit is reverted, can be skipped.

outdoor2kode committed:
Make plugin unable to be network deactivated.

outdoor2kode committed:
Only delete temp attendee when it's next-gen

outdoor2kode commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/cron.php:
```diff
@@ -0,0 +1,200 @@
+function add_temp_attendee() {
+	if ( ! get_wordcamp_attendees_id() ) {
+		$attendees_id = wp_insert_post( array(
+			'post_title' => 'Organizer debrief survey',
```
![image](https://github.com/WordPress/wordcamp.org/assets/18050944/a9a5171e-5d5b-4ea6-9381-c9d1aeb405b0)
![image](https://github.com/WordPress/wordcamp.org/assets/18050944/cfe52659-c476-4278-8e7d-0e509e40d53d)


The title is `(no title)` even though the `post_title` is specified and updated to DB. And after moving it to the Trash, the Title appears. I'm not entirely sure why this happens, but it doesn't affect the overall process.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/wordcamp-organizer-survey.php:
```diff
@@ -0,0 +1,151 @@
+
+	// Flushing the rewrite rules is buggy in the context of `switch_to_blog`.
+	// The rules will automatically get recreated on the next request to the site.
+	delete_option( 'rewrite_rules' );
```
Since this function is hooked to `init`, this will run on every page load. It's very expensive, so it's best to only do it during the activation hook.

Is it necessary to call it, though? I don't see any rewrite rules being added.

If it is needed, then `flush_rewrite_rules()` could be used directly, since this isn't being called during a `switch_to_blog()` process.

outdoor2kode replied:
Ah, yeah. You're right. This should be removed. Previously, when I tried to add the Organizer Reminder, I used `switch_to_blog`, so it's been there ever since.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/admin-page.php:
```diff
@@ -0,0 +1,145 @@
+	$feedback_details = array();
+
+	foreach ( $wordcamps as $camp ) {
+		// It's a bit counter intuitive, but the site ID is the blog ID.
```
This is a legacy inconsistency that stems from Core, where it was originally called a "blog", but then later called a "site". These days it's always called a site, but some of the old things like database schemas still say "blog" due to back-compat

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/page.php:
```diff
@@ -0,0 +1,341 @@
+		return;
+	}
+
+	$page_id = wp_insert_post( array(
```
🤔 I wonder if it's better to create a single page on Central, rather than a new page on each camp site? There'll only be 1 respondent per camp, so it seems it might be overkill to do it on every site.

It could remove the need for `admin-page.php`, code to create/publish/delete/protect the page, etc.

I don't feel strongly, though. What do you think?

outdoor2kode replied:
> I wonder if it's better to create a single page on Central, rather than a new page on each camp site? There'll only be 1 respondent per camp, so it seems it might be overkill to do it on every site.

If we only create a page on the central site, then we can't have the mechanism to set this page as a draft to prevent spam. 

We currently have an anti-spam mechanism where, two weeks after the survey page link sent to organizer, this published page is automatically deleted to prevent organizers from accessing it. 

If we were to create a page on the central site, the timing of changing the page status from draft to publish (add the page to camptix email) triggered by one site would overlap with the anti-spam mechanism's change from publish to draft triggered by another site. 

That said, given that only one person, the lead organizer, is filling out the survey for each event, the chances of spam, while still possible, are considerably reduced.

Perhaps in this scenario, we don't need to worry about preventing spam?

> It could remove the need for admin-page.php, code to create/publish/delete/protect the page, etc.

I understand we can remove the need for code to create/publish/delete/protect the page if moving the page to central, but I'm a bit not sure how this can also affect the need for admin-page.php.

iandunn replied:
> then we can't have the mechanism to set this page as a draft to prevent spam.

🤔 That's a good point. We could manually create the page on Central, and leave it published all the time. We then might be able to require the organizer to log in to submit an entry, and we could automatically populate the "which wordcamp..." field based on that. We could show them an error if their camp was over more than 2 weeks ago, or maybe we wouldn't have to worry about that because we know only the organizer can fill it in.

iandunn replied:
>  how this can also affect the need for admin-page.php.

I was thinking that it wouldn't be needed, since admins could use a workflow like this:

1. Visit https://central.wordcamp.org/wp-admin/edit.php?post_type=feedback
2. change `All sources` to the specific form
3. click `Filter` or `Export`

iandunn replied:
🤔 Rather than requiring them to log in, we could also just put a token in the URL that we send them, and use that to verify that they're an organizer. That might avoid some folks skipping it because of the extra work of logging in.

It may also avoid some edge cases where there's extra hassles. For example, we store the email address of the lead organizer, but not their w.org username. So if they log in but their w.org account email doesn't match the email we sent the request to, then they might have to update the email to match.

For the token, we might be able to just create a secret key in the config, and then create a hash of their email address and the key; something like `hash_hmac( 'sha1', $email, ORGANIZER_SURVEY_ACCESS_TOKEN_KEY )`. That way we wouldn't have to store any additional metadata in the database. If having the value in the db is needed for other parts of the code, though, then that's not a problem either. In that case we could probably just use `wp_generate_password()` to create the token.

outdoor2kode replied:
Thanks for the suggestion 👍 👍 👍 
I've removed the admin page and moved the survey page to central.

And after talking a bit with Steve earlier today, we figure it seems we can get more things on the central site if including the `wordcamp_id` in the URL - capable of answering [as many questions as possible](https://github.com/WordPress/wordcamp.org/pull/1038#discussion_r1329418934) with the event site data retrieved from it. 
So it would be `hash_hmac( 'sha1', $wordcamp_id, ORGANIZER_SURVEY_ACCESS_TOKEN_KEY )`.

The changes starts from this commit [7354b3a](https://github.com/WordPress/wordcamp.org/pull/1038/commits/7354b3a7d7e13dc43a3a2b7f65cf34810e05ef21)

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/cron.php:
```diff
@@ -0,0 +1,200 @@
+}
+
+/**
+ * Add temporary attendee since email can only be sent to items in attendee tracker.
```
This feels like a bit of a hack. I wonder if there's a way to make the CampTix email more flexible, so that it doesn't require an attendee?

Or maybe this plugin shouldn't be so tightly coupled with CampTix?

I can see how it's a practical solution, though, so it may not be worth changing. What do you think?

outdoor2kode replied:
> Or maybe this plugin shouldn't be so tightly coupled with CampTix?

When I was working on this PR previously, I had intended to use the Organizer Reminder to address it. However, my initial approach was in the totally wrong direction, leading me to spend an excessive amount of time on it. That's why I eventually resorted to this hacky but practical method.

Today I spent another day trying to see if the Organizer Reminder could potentially resolve this issue. I believe it's feasible, but the main hurdle is the spam concern mentioned above. If we can bypass the need to set the page to draft for spam prevention, then based on how Organizer Reminder operates, the solution would be: setting up an Automated Job with the email content prepared, including the link to the published survey page. It would then scan the WordCamp posts on a daily basis to identify any WordCamp events that meet the sending-reminder criteria, fetch their data, include it in the context of the automated job, and send the email.

What do you think of this approach?

outdoor2kode replied:
> I can see how it's a practical solution, though, so it may not be worth changing.

If we proceed with merging the current PR, as I've made some preparations here to generalize the plugin, we can open another PR to integrate the Attendee Survey, integrating it into a single Survey Plugin.

iandunn replied:
That approach with Organizer Reminders sounds good 👍🏻 I also think it's reasonable to stick with this approach if you want. Organizer Reminders may be more ideal, but it also may take more time, so it's a trade off. If we'll want to refactor this later, though, then that also takes time 🤷🏻‍♂️ 

I don't have a strong opinion either way, but I trust your judgement 👍🏻

StevenDufresne replied:
Let's prioritize merging this over architecture changes. This feature hasn't proved its worth yet.

outdoor2kode replied:
Thanks for providing extra insights on the path moving forward @iandunn and @StevenDufresne, I'll proceed with the current approach.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/cron.php:
```diff
@@ -0,0 +1,200 @@
+
+	$end_date = $wordcamp_post->meta['End Date (YYYY-mm-dd)'][0];
+
+	if ( ! isset( $end_date ) ) {
```
IIRC, some sites that are only 1-day events will only have the start date. If the end date is empty and the start date is filled in, then the start date can be used as the end date.

outdoor2kode replied:
Good point, thanks!

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/page.php:
```diff
@@ -0,0 +1,341 @@
+	
+	<!-- wp:jetpack/contact-form {"subject":"New feedback received from your website","style":{"spacing":{"padding":{"top":"16px","right":"16px","bottom":"16px","left":"16px"}}}} -->
+	<div class="wp-block-jetpack-contact-form" style="padding-top:16px;padding-right:16px;padding-bottom:16px;padding-left:16px"><!-- wp:group {"layout":{"type":"constrained"}} -->
+	<div class="wp-block-group"><!-- wp:jetpack/field-text {"label":"What event did you organize?","required":true,"requiredText":"(required)"} /--></div>
```
We shouldn't need to ask this, since we already know it based on the current site, who the email is sent to, etc. We can remove the question, but have some code that automatically adds it to the response before it's saved in the database.

outdoor2kode replied:
It seems there's no proper filter or approach to add the predefined data before inserting the response to Feedback for [Jetpack contact form](https://github.com/codeforamerica/cityhow/blob/master/wp-content/plugins/jetpack/modules/contact-form/grunion-contact-form.php#L1038). What's the common way to do so?

Also tried `grunion_pre_message_sent` and `wp_insert_post_data`, but I feel they aren't the best solution here.

Maybe using JavaScript to set the default values? I feel like there's a more elegant way to do so.

outdoor2kode replied:
Known data was added in ad40047c6cf40e48ce4c82170b023d0bc2666690

iandunn replied:
Yeah, I wish there were more filters there 😦 

It seems like it might work to hook into `template_redirect` at priority `5`, since `process_form_submission()` runs at priority `10`. Then you might be able to modify `$_POST` directly.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/page.php:
```diff
@@ -0,0 +1,341 @@
+	<!-- /wp:heading -->
+	
+	<!-- wp:group {"layout":{"type":"constrained"}} -->
+	<div class="wp-block-group"><!-- wp:jetpack/field-text {"label":"How many tickets were sold?","required":true,"requiredText":"(required)"} /--></div>
```
This can also be pulled from existing data. There may be other questions that can also be removed for that reason.

outdoor2kode replied:
In this PR, I've pre-populated the value for the question "What event did you organize?". The data sources for the other questions are different and might need to look for proper functions to retrieve the existing data. I'll handle that in a separate PR.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/cron.php:
```diff
@@ -0,0 +1,200 @@
+	if ( ! $result ) {
+		log( 'Failed to add recipients:', $email_id, $recipient_id );
+	} else {
+		update_post_meta( $email_id, 'tix_email_recipients_backup', (array) $recipient_id );
```
Did you notice this happening, or is it more to protect against a theoretical situation?

outdoor2kode replied:
Yes, I noticed there's a warning
```
Warning: count(): Parameter must be an array or an object that implements Countable in /usr/src/public_html/wp-content/plugins/camptix/camptix.php on line 808
```
<img width="1534" alt="image" src="https://github.com/WordPress/wordcamp.org/assets/18050944/9a5b2875-70eb-4a16-92d4-7f7fc0fd79fa">


---

After investigating the issue, I discovered that this line was returning an empty string for recipients_backup
https://github.com/WordPress/wordcamp.org/blob/ea44ddd8008b565b3b32da99be3149aacad12d38/public_html/wp-content/plugins/camptix/camptix.php#L807
This would make it counted as `1`. I've pushed ea44ddd8008b565b3b32da99be3149aacad12d38 and 0510637a93f98ad6950eab47cb975c737b4d0426 to fix it.

iandunn reviewed (approved):
I left a few suggestions but this looks good overall

outdoor2kode committed:
Remove delete_option( 'rewrite_rules' );

outdoor2kode committed:
Set end_date as start_date for 1-day events.

outdoor2kode committed:
Fix tix_email_recipients_backup wrongly updated

outdoor2kode committed:
Remove debrief page.

outdoor2kode committed:
Remove Admin Page.

outdoor2kode committed:
Embed a token in URL for verifying organizer.

outdoor2kode committed:
Update email content

outdoor2kode committed:
Format Cleanup

outdoor2kode committed:
Encrypt with wordcamp_id rather than email.

outdoor2kode committed:
Fill in data that we already know.

outdoor2kode committed:
Fix linting error.

iandunn commented on .docker/wp-config.php:
```diff
@@ -118,6 +118,8 @@
 define( 'EMAIL_DEVELOPER_NOTIFICATIONS', 'developers@example.test' );
 define( 'EMAIL_CENTRAL_SUPPORT',         'support@wordcamp.test' );
 
+define( 'ORGANIZER_SURVEY_ACCESS_TOKEN_KEY', 'wcorg_debrief_survey_email' );
```
```suggestion
define( 'ORGANIZER_SURVEY_ACCESS_TOKEN_KEY', 'put your unique phrase here' );
```

This might make it more obvious that the value is essentially a password, so it should be a long randomly generated  value, just like the key/salt values above.

iandunn commented on public_html/wp-content/mu-plugins/wcorg-network-plugin-control.php:
```diff
@@ -94,6 +94,7 @@ function _get_network_plugin_state_list( $state ) {
 	} elseif ( EVENTS_NETWORK_ID === $network_id ) {
 		$network_plugin_state['deactivated'][] = 'tagregator/bootstrap.php';
 		$network_plugin_state['activated'][]   = 'camptix-attendee-survey/camptix-attendee-survey.php';
+		// $network_plugin_state['activated'][]   = 'wordcamp-organizer-survey/wordcamp-organizer-survey.php';
```
Should this be commented out?

outdoor2kode replied:
Ah, thanks for catching it!

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/wordcamp-organizer-survey.php:
```diff
@@ -0,0 +1,178 @@
+
+		$expected_token = hash_hmac( 'sha1', base64_decode( $wordcamp_id ), ORGANIZER_SURVEY_ACCESS_TOKEN_KEY );
+
+		if ( $token !== $expected_token ) {
```
```suggestion
		if ( $token !== $expected_token ) {
		if ( ! hash_equals( $expected_token, $token ) {
```

Using `hash_equals()` avoids [timing attacks](https://security.stackexchange.com/a/83670/8467). It's not a big deal in this scenario, but it's a good habit to get into, and might get flagged by a linter in the future if we don't.

outdoor2kode replied:
TIL 👍 👍 👍

outdoor2kode committed:
Disable validation and automatic filling out.

The Jetpack contact form fails to submit when there's a query string present.The Jetpack contact form's source code has been traced and no usable hooks found. Although the query string can be removed via JavaScript, the form still can't be submitted successfully. Therefore, it's temporarily disabled.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/wordcamp-organizer-survey.php:
```diff
@@ -0,0 +1,178 @@
+		$labels = $dom->getElementsByTagName('label');
+		foreach ( $labels as $label ) {
+			// phpcs:disable WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+			if ( 'What event did you organize?' === $label->nodeValue ) {
```
🤔 I suspect there may an easier way to do this. Do we need to show the question to the user if we're just going to fill it in anyway? 

Another problem is that, If it's in the page, it could be edited by an admin who wouldn't realize that doing that would break this.

If we don't need to show the question, we could insert the value in `$_POST` some time before `Grunion_Contact_Form_Plugin::process_form_submission()` runs.

If you do want to show it, it might be possible to add an `id/class` value to the post content in the database, and then use that in this function. That'd still break if someone edited it, but they'd be less likely to.

outdoor2kode committed:
Fix email stats counting error.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/wordcamp-organizer-survey.php:
```diff
@@ -0,0 +1,178 @@
+			if ( 'What event did you organize?' === $label->nodeValue ) {
+				$input_id         = $label->getAttribute('for');
+				$event_name_field = $dom->getElementById($input_id);
+				$event_name_field->setAttribute('value', $wordcamp_post_data->post_title);
```
I think it'd be best to save the ID rather than the title, since the title can change. It also won't have the year, unless passed through `get_wordcamp_name()`. That function might not work well for NextGen camps though, since some of them are monthly, etc.

If we don't remove the question from the user-facing form (see above), then it would be good to show the user the title, but we could still save the ID. Maybe just by modifying `$_POST` before the value gets saved. Or it could be in a separate field if that's easier.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/cron.php:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	// Remove the cron job that queues everything.
+	wp_clear_scheduled_hook( 'wc_organizer_debrief_survey_email' );
```
It looks like this will just get added again on the next page load by `Cron\schedule_jobs()`, is that right? if so, is there any advantage to removing it here?

outdoor2kode committed:
Reveert: Disable validation and automatic filling out and fix the issue

outdoor2kode committed:
Text field can't be disabled or jetpack unable to get the data.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/email.php:
```diff
@@ -0,0 +1,148 @@
+ *
+ * @return string
+ */
+function get_encryption_token() {
```
```suggestion
function get_verification_token() {
```

Technically this is [hashing, rather than encryption](https://www.thesslstore.com/blog/difference-encryption-hashing-salting/).

outdoor2kode replied:
Ah, yes, you're right. Messed up on the basics.. Thanks!

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/email.php:
```diff
@@ -0,0 +1,148 @@
+ */
+function get_encryption_token() {
+	$wordcamp_id = get_current_blog_id();
+	return hash_hmac( 'sha1', $wordcamp_id, ORGANIZER_SURVEY_ACCESS_TOKEN_KEY );
```
```suggestion
	return hash_hmac( 'sha256', $wordcamp_id, ORGANIZER_SURVEY_ACCESS_TOKEN_KEY );
```

I should have double checked if `sha1` was still considered secure before suggesting it; turns out it's not. `sha256` should be fine though.

Similar to `hash_equals()`, it's not really critical for this application, but a good habit to get into.

The other `sha1` instance will need to be updated to match.

outdoor2kode replied:
Actually I did consider the fact that `sha1` can be compromised initially. I just thought later on that maybe there was a reason you suggested `sha1`, like perhaps it doesn't need such a high level of security for this particular use case here, and `sha1` performs better in terms of performance 😆

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/email.php:
```diff
@@ -0,0 +1,148 @@
+function get_email_content() {
+	$wordcamp_name   = get_wordcamp_name();
+	$survey_page_url = 'https://central.wordcamp.test/organizer-survey-event-debrief/?t=' . get_encryption_token()
+					 . '&wid=' . base64_encode( get_current_blog_id() );
```
What's the advantage of using `base64_encode()` here?

outdoor2kode replied:
Because originally, the email was placed here. And with considerations for when users open their emails or click on the link, adding extra protection to ensure others can't easily see the user's email from the screen. I think we can remove it here. Even if the `wordcamp_id` is seen, it's not easy to misuse it.

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/email.php:
```diff
@@ -0,0 +1,148 @@
+ */
+function get_email_content() {
+	$wordcamp_name   = get_wordcamp_name();
+	$survey_page_url = 'https://central.wordcamp.test/organizer-survey-event-debrief/?t=' . get_encryption_token()
```
```suggestion
	$tld = get_top_level_domain();
	$survey_page_url = "https://central.wordcamp.$tld/organizer-survey-event-debrief/?t=" . get_encryption_token()
```

outdoor2kode replied:
Ah wasn't aware there's already a function like this. Was planning to handle that before shipping to production. 👍

iandunn commented on public_html/wp-content/plugins/wordcamp-organizer-survey/includes/debrief-survey/email.php:
```diff
@@ -0,0 +1,148 @@
+	$email .= "<strong>Event Recording (if any)</strong>\r\n";
+	$email .= sprintf( "If you haven't yet done so, please review the submission guidelines before beginning to edit your videos (or before your videographers starts editing): %s\r\n\r\n", esc_url( 'http://blog.wordpress.tv/submission-guidelines/' ) );
+	$email .= sprintf( "To submit your video for publication to WordPress.tv, just upload them at this page: %s\r\n\r\n", esc_url( 'http://wordpress.tv/submit-video/' ) );
+	$email .= 'Our intrepid team of video moderators will review the videos and schedule them for publication. Our intrepid team of video moderators will review the videos and schedule them for publication. ';
```
```suggestion
	$email .= 'Our intrepid team of video moderators will review the videos and schedule them for publication.';
```

iandunn reviewed (approved):
I left some questions/suggestions, but nothing that blocks merging 👍🏻

outdoor2kode committed:
Update ORGANIZER_SURVEY_ACCESS_TOKEN_KEY.

outdoor2kode committed:
Comment out network_plugin_state.

outdoor2kode committed:
Using hash_equals() to avoid timing attacks.

outdoor2kode committed:
Correctly rename get_verification_token.

outdoor2kode committed:
Update to use sha256.

outdoor2kode committed:
Remove base64 encoding and decoding.

outdoor2kode committed:
Get top level domain with built-in function.

outdoor2kode committed:
Remove copy duplication.
